openedx_learning/core/publishing/api.py

```diff
@@ -0,0 +1,192 @@
+        key=key,
+        title=title,
+        created=created,
+        updated=created,
+    )
+    package.full_clean()
+    package.save()
+
+    return package
+
+
+def create_publishable_entity(learning_package_id, key, created, created_by):
+    """
+    Create a PublishableEntity.
+
+    You'd typically want to call this right before creating your own content
+    model that points to it.
+    """
+    return PublishableEntity.objects.create(
+        learning_package_id=learning_package_id,
+        key=key,
+        created=created,
+        created_by=created_by,
+    )
+
+
+def create_publishable_entity_version(
+    entity_id, version_num, title, created, created_by
+):
+    """
+    Create a PublishableEntityVersion.
+
+    You'd typically want to call this right before creating your own content
+    version model that points to it.
+    """
+    with atomic():
+        version = PublishableEntityVersion.objects.create(
+            entity_id=entity_id,
+            version_num=version_num,
+            title=title,
+            created=created,
+            created_by=created_by,
+        )
+        Draft.objects.create(
+            entity_id=entity_id,
+            version=version,
+        )
+    return version
+
+
+def learning_package_exists(key: str) -> bool:
+    """
+    Check whether a LearningPackage with a particular key exists.
+    """
+    LearningPackage.objects.filter(key=key).exists()
+
+
+def publish_all_drafts(
+    learning_package_id, message="", published_at=None, published_by=None
+):
+    """
+    Publish everything that is a Draft and is not already published.
+    """
+    draft_qset = (
+        Draft.objects.select_related("entity__published")
+        .filter(entity__learning_package_id=learning_package_id)
+        .exclude(entity__published__version_id=F("version_id"))
+    )
+    return publish_from_drafts(
+        learning_package_id, draft_qset, message, published_at, published_by
+    )
+
+
+def publish_from_drafts(
+    learning_package_id: int,  # LearningPackage.id
+    draft_qset: QuerySet,
+    message: str = "",
+    published_at: Optional[datetime] = None,
+    published_by: Optional[int] = None,  # User.id
+) -> PublishLog:
+    """
+    Publish the rows in the ``draft_model_qsets`` args passed in.
+    """
+    if published_at is None:
+        published_at = datetime.now(tz=timezone.utc)
+
+    with atomic():
+        # One PublishLog for this entire publish operation.
+        publish_log = PublishLog(
+            learning_package_id=learning_package_id,
+            message=message,
+            published_at=published_at,
+            published_by=published_by,
+        )
+        publish_log.full_clean()
+        publish_log.save(force_insert=True)
+
+        for draft in draft_qset.select_related("entity__published"):
+            # There might not be a currently published version for this
+            # PublishableEntity yet.
+            if hasattr(draft.entity, "published"):
```

Comment:
@ormsbee sorry, I'm confused about this too: how does `published` get assigned on the PublishableEntity instance? Why is it not always available on the instance? I would have expected this check to look like:
```python
if draft.entity.published:
```

Reply:
The `Published` model has the OneToOneField pointing to `PublishableEntity`. So when you're coming through the reverse relation (`entity.published`), there's a chance that there is no row in `Published`. When that happens, Django will throw a `ObjectDoesNotExist` if you try to access it, and they recommend checking with `hasattr`:

https://docs.djangoproject.com/en/3.2/topics/db/examples/one_to_one/

Reply:
Done.